R/utils.R

```diff
@@ -58,40 +58,63 @@ py_global_env <- function() {
   reticulate::py
 }
 
-# Create a duplicate of a Python environment
-#
-# @examples
-# reticulate::py_run_string("x = 3")
-# new_py_envir <- duplicate_py_env(py)
-# new_py_envir$items()
-#
-# @param module the `py` module accessed via `reticulate`
-#
-# @return a Python `Dict` or dictionary that is not converted to an R data type
-# @keywords internal
-duplicate_py_env <- function(module) {
+py_global_dict <- function() {
+  reticulate::py_get_attr(py_global_env(), "__dict__")
+}
+
+#' Create a duplicate of a Python environment
+#'
+#' @return a Python `Dict` or dictionary
+#'
+#' @keywords internal
+#'
+#' @examples
+#' \dontrun{
+#' reticulate::py_run_string("x = 3")
+#' new_py_envir <- py_copy_global_env()
+#' new_py_envir$items()
+#' }
+py_copy_global_env <- function() {
   rlang::check_installed("reticulate", "Python exercise support")
 
-  # extract all objects within this module
-  new_objs <- reticulate::py_get_attr(module, "__dict__")
-  # then create copy of the dictionary with all objects
-  copy <- reticulate::import("copy", convert = FALSE)
-  copy$copy(new_objs)
-}
-
-# This clears the Python environment `py`.
-#
-# It will keep important initial objects such as `py` (main module), 
-# `r` (reticulate interface to R), and the `builtins` module.
-#
-# @examples
-# reticulate::py_run_string("x = 3")
-# # this removes the `x`
-# clear_py_env()
-#
-# @return Nothing
-# @keywords internal
-clear_py_env <- function() {
+  # grab Python utility functions
+  py_utils <- py_learnr_utilities()
+
+  # extract all objects of `reticulate::py` (the main module)
+  py_env_dict <- py_global_dict()
+
+  # deep copy Python environment
+  reticulate::py_call(py_utils$deep_copy, py_env_dict)
```

Comment:
```suggestion
  py_utils <- py_learnr_utilities()

  reticulate::py_call(py_utils$deep_copy, py_global_dict())
```

I think this is readable enough that we don't need the comments, although I would suggest adding a comment about why we need to use `py_call()` here.

Reply:
Yeah agreed!